taskqueue/gcloud/aio/taskqueue/taskmanager.py

```diff
@@ -2,486 +2,159 @@
-                    'flunked': self.flunked_tasks,
-                    'failed': self.failed,
-                    'progressing': self.no_tasks,
-                    'rolling_duration': self.durations.avg,
-                    'rolling_size': self.sizes.avg
-                },
-                'created_at': str(self.created_at),
-                'state': getattr(self.state, 'name', None)
-            }
+        properties = {
+            'bucket': payload.get('bucket'),
+            'error': str(exception),
+            'generation': None,
+            'metageneration': None,
+            'time_created': datetime.datetime.now(),
+            'traceback': traceback.format_exc(exception),
+            'update': None,
         }
 
-    async def print_stats(self):
-
-        stats = await self.get_stats()
-        log.info(stats)
-
-        call_later(10, self.print_stats)
+        await self.deadletter_insert_function(task['name'], properties)
 
     async def poll(self):
-
-        """
-        poll a task queue for work at self.poll_interval
-        """
-
-        state = None
-        args = None
-
-        default_step = self.try_to_lease
-        state_step = {'LEASE': self.after_lease}
-        stepper = make_stepper(default_step, state_step, name='TM')
-
-        # TODO: migrate away from stepper + AwaitableState
         while self.running:
-            state, args = await stepper(state, args)
-            self.state = state
-
-    def after_lease(self, tasks):
-
-        if not tasks:
-
-            interval = self.poll_interval + next(self.backoff)
-
-            return AwaitableState(
-                'NO_TASKS',
-                asyncio.sleep(interval)
-            )
-
-        # reset the backoff
-        self.backoff = backoff(base=1.1, max_value=self.max_poll_interval)
+            # only lease new tasks when we have room
+            async with self.semaphore:
+                task_lease = await self.tq.lease(
+                    lease_seconds=self.lease_seconds,
+                    num_tasks=self.batch_size)
 
-        for task in tasks:
+            if not task_lease:
+                await asyncio.sleep(next(self.backoff))
+                continue
 
-            payload = deserialize_task(task)
-            self.process_task(task, payload)
+            tasks = task_lease.get('tasks')
+            log.info('grabbed %d tasks', len(tasks))
 
-        return AwaitableState(
-            'PROCESS',
-            asyncio.sleep(self.poll_interval),
-            data=len(tasks)
-        )
+            futures = []
+            for task in tasks:
+                futures.append(asyncio.ensure_future(self.process(task)))
 
-    def try_to_lease(self):
+            await asyncio.gather(*futures)
 
-        if self.no_tasks >= self.max_concurrency:
+    async def process(self, task):
+        name = task['name']
+        payload = ujson.loads(decode(task['pullMessage']['payload']))
```

Comment:
Did you mention that the payload isn't always a JSON object?

Reply:
Adding a comment.